src/mono/mono/utils/mono-time.c

```diff
@@ -229,3 +235,84 @@ mono_100ns_datetime_from_timeval (struct timeval tv)
 
 #endif
 
+#if defined(HOST_DARWIN) || defined(HOST_LINUX)
+
+void
+mono_clock_init (void *clk_id)
+{
+#ifdef HOST_DARWIN
+	kern_return_t ret;
+
+	do {
+		ret = host_get_clock_service (mach_host_self (), SYSTEM_CLOCK, (clock_serv_t *)clk_id);
+	} while (ret == KERN_ABORTED);
+
+	if (ret != KERN_SUCCESS)
+		g_error ("%s: host_get_clock_service () returned %d", __func__, ret);
+#endif	
+}
+
+void
+mono_clock_cleanup (void *clk_id)
+{
+#ifdef HOST_DARWIN
+	kern_return_t ret;
+
+	do {
+		ret = mach_port_deallocate (mach_task_self (), *(clock_serv_t *)clk_id);
+	} while (ret == KERN_ABORTED);
+
+	if (ret != KERN_SUCCESS)
+		g_error ("%s: mach_port_deallocate () returned %d", __func__, ret);
+#endif
+}
+
+guint64
+mono_clock_get_time_ns (void *clk_id)
```

Comment:
Why do you need to pass an address to a clock ID, just to dereference it here in the function?  Couldn't you just pass a constant?

Reply:
Because it is not a constant for type `clock_serv_t` on macOS
`typedef struct clock *clock_serv_t;`